File: ProblemSolving/450DSA/JavaScript/src/dynamicprogramming/SpaceOptimizedLongestCommonSubsequence.js
/**
 * @author Anirudh Sharma
 * <p>
 * Given two strings, find the length of longest subsequence present in both of them.
 * Can we do this in O(n) space
 */
const spaceOptimizedLCS = (s1, s2) => {
    // Special case
    if (s1 === undefined || s1.length === 0 || s2 === undefined || s2.length == 0) {
        return 0;
    }
    // Lengths of two strings
    const m = s1.length;
    const n = s2.length;
    // Lookup table to store the longest common subsequence
    // for a given value of m and n.
    const lookup = Array.from(Array(2), () => Array(n + 1).fill(0));
    // Binary index used to index current and previous rows
    let binary = 0;
    // Loop over the strings
    for (let i = 1; i <= m; i++) {
        // Current binary index
        binary = i & 1;
        for (let j = 1; j <= n; j++) {
            if (s1.charAt(i - 1) === s2.charAt(j - 1)) {
                lookup[binary][j] = 1 + lookup[1 - binary][j - 1];
            } else {
                lookup[binary][j] = Math.max(lookup[1 - binary][j], lookup[binary][j - 1]);
            }
        }
    }
    return lookup[binary][n];
};

const main = () => {
    let s1 = "ABCDGH";
    let s2 = "AEDFHR";
    console.log(spaceOptimizedLCS(s1, s2));

    s1 = "ABC";
    s2 = "AC";
    console.log(spaceOptimizedLCS(s1, s2));
};

main();

File: ProblemSolving/450DSA/Python/src/dynamicprogramming/SpaceOptimizedLongestCommonSubsequence.py
"""
 @author Anirudh Sharma
 
Given two strings, find the length of longest subsequence present in both of them.
Can we do this in O(n) space?
"""


def spaceOptimizedLCS(s1, s2):
    # Special case
    if s1 is None or len(s1) == 0 or s2 is None or len(s2) == 0:
        return 0
    # Lengths of the two strings
    m, n = len(s1), len(s2)
    # Lookup table to store the longest common subsequence
    # for a given value of m and n.
    lookup = [[0 for y in range(n + 1)] for x in range(2)]
    # Populate the table for each character
    for i in range(1, m + 1):
        binary = 1 & i
        for j in range(1, n + 1):
            # If current characters are same
            if s1[i - 1] == s2[j - 1]:
                lookup[binary][j] = 1 + lookup[1 - binary][j - 1]
            else:
                lookup[binary][j] = max(
                    lookup[1 - binary][j], lookup[binary][j - 1])
    return lookup[binary][n]


if __name__ == "__main__":
    s1 = "ABCDGH"
    s2 = "AEDFHR"
    print(spaceOptimizedLCS(s1, s2))

    s1 = "ABC"
    s2 = "AC"
    print(spaceOptimizedLCS(s1, s2))
